web/frontend/src/components/slam-2d-render.vue

```diff
@@ -60,54 +77,80 @@ const disposeScene = () => {
   scene.clear();
 };
 
-const update = (pointcloud: Uint8Array, pose: commonApi.Pose) => {
+const updateCloud = (pointcloud: Uint8Array) => {
+  disposeScene();
+
+  const viewHeight = 1;
+  const viewWidth = viewHeight * 2;
+
   const points = loader.parse(pointcloud.buffer, '');
+  points.geometry.computeBoundingSphere();
 
-  const x = pose.getX!();
-  const z = pose.getZ!();
-  marker.position.setX(x);
-  marker.position.setZ(z);
+  const { radius = 1, center = { x: 0, z: 0 } } = points.geometry.boundingSphere ?? {};
+  camera.position.set(center.x, 100, center.z);
+  camera.lookAt(center.x, 0, center.z);
+
+  const aspect = canvas.clientHeight / canvas.clientWidth;
+  camera.zoom = aspect > 1
+    ? viewHeight / (radius * 2)
+    : camera.zoom = viewWidth / (radius * 2);
+
+  camera.updateProjectionMatrix();
+
+  controls.target.set(center.x, 0, center.z);
+  controls.maxZoom = radius * 2;
+
+  const intersectionPlane = new THREE.Mesh(
+    new THREE.PlaneGeometry(radius * 2, radius * 2, 1, 1).rotateX(-Math.PI / 2),
+    new MeshDiscardMaterial()
+  );
+  intersectionPlane.name = 'Intersection Plane';
+  intersectionPlane.position.y = -1;
+  intersectionPlane.position.set(center.x, 0, center.z);
+  raycaster.objects = [intersectionPlane];
 
-  disposeScene();
   scene.add(points);
   scene.add(marker);
+  scene.add(intersectionPlane);
 };
 
-const init = (pointcloud: Uint8Array, pose: commonApi.Pose) => {
-  update(pointcloud, pose);
+const updatePose = (newPose: commonApi.Pose) => {
+  const x = newPose.getX();
+  const z = newPose.getZ();
+  marker.position.setX(x);
+  marker.position.setZ(z);
 };
 
 onMounted(() => {
-  container.append(renderer.domElement);
-
-  camera.position.set(0, 100, 0);
-  camera.lookAt(0, 0, 0);
+  container?.append(canvas);
 
-  renderer.setAnimationLoop(() => {
-    resizeRendererToDisplaySize(camera, renderer);
+  run();
 
-    renderer.render(scene, camera);
-    controls.update();
-  });
+  if (props.pointcloud !== undefined) {
```

Comment:
This is a small optimization. Pose updates and cloud updates are decoupled. We shouldn't break down the entire scene, parse the entire PCD buffer, and generate the bounding sphere every time pose changes.